### PR TITLE
BifunctorTrans type class

### PR DESCRIPTION
**Pull request by jystic**
I'm not 100% sure about this one, so I wanted to get some opinions before I go and write tests for it.

Essentially this is `Bifunctor` but for monad transformers. Monad transformers can't implement `Bifunctor` because we can't rearrange the type parameters in a syntactically friendly way.

The main motivation is that we use `bimapEitherT` and `firstEitherT` a lot, and they're too damn long :)

### CODE REVIEW

**Comment by tranma** [comment]
:+1: in hipchat space

**Comment by tmcgilchrist** [comment]
```haskell
bimapEitherT :: Functor m => (e -> f) -> (a -> b) -> EitherT e m a -> EitherT f m b

# vs

bimapT :: Functor f => (x -> y) -> (a -> b) -> t x f a -> t y f b
```

**Comment by jystic** [comment]
```haskell
firstEitherT :: Functor m => (e -> f) -> EitherT e m a -> EitherT f m a

# vs

firstT :: Functor f => (x -> y) -> t x f a -> t y f a
```

**Comment by domdere** [comment]
I'm ok with this existing as a concept, `Bifunctor` is enough to justify this as a "Thing".  The only downside i can see is I dont see it getting the same mileage as `Bifunctor` in our codebase.  The only instance i can think of actually using is the `ExceptT/EitherT` instance, which we already have `bimapEitherT` for, even though its more characters. `WriterT` doesnt get much use.

While you can say the same thing about `Bifunctor` (i.e its real only use case apart from all constructs in the internals of Lens is `Either`), `Bifunctor` gets a lot of use because we define a lot of sum types with two covariant type parameters that are similar to `Either`.

With `BifunctorTrans`, we would exercise it more if we defined a lot of monad transformers ourselves that were effectively `ExceptT`, which is just something thats not commonplace in our style we usually just use `ExceptT`.

It *might* come in handy when dealing with third party stacks that are parameterised on the error type (e.g `WebMachine` (? not sure off the top of my head whether this is parameterised on the error type)).  Maybe @tmcgilchrist or @charleso can comment on whether `WebMachine` would provide an example use case this.

Anyway not against it perse, don't see the use with the codebase as it stands, but I do try and be sympathetic when ppl try to reduce the amount of characters they have to type, as long as it does no harm and i dont think it does any harm here.

**Comment by markhibberd** [comment]
@jystic I am good with it based on what @domdere says. My only question is how we deal with instances. The problem being that if there are other third party libraries which can/should have instances where do they go? Would be nice to have an answer before we merge, even if it is just a vague idea of what we would do (perhaps `x-bifunctor-trans-<package>` orphans, or resurrect the `orphanarium` project for more general usage). I just really don't want to end up with orphan's littered over the code bases.

**Comment by jystic** [comment]
@domdere I could imagine a combined Reader/State/Either transformer in the same vain as `RWST` that could make use of it, but as you say we usually just use `ExceptT`.

@markhibberd I think the question about orphans would be good to sort out in general, as `NFData` and `Typeable` have the same problem sometimes.

I'd be ok with just putting them in `x-<package>` in a separate module, are there any problems with that? If so perhaps `orphanarium-<package>` is better.

**Comment by markhibberd** [comment]
@jystic I think I would err on the side of a dedicated spot so it doesn't cloud the rules. Create a dedicated `orphanarium` project, which has fine grained `orphanarium-<package>` sub-projects. This will help answer some of the questions @olorin was asking. I just want the rules to be really clear, there are a few orphans that can't be avoided as you mentioned, but other than that I have enough scars to never want to see an orphan for any reason :)

**Comment by jystic** [comment]
@markhibberd agreed, I'm +1 for `orphanarium-<package>`

**Comment by jystic** [comment]
Tests added /cc @markhibberd @domdere

File: test/Test/P/Bifunctor/Trans.hs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  -> Fun x y
+  -> Fun a b
+  -> Property
+bifunctorLaws mkT runT p (Fun _ f) (Fun _ g) =
+  conjoin [
+      functionEq mkT runT p (bimapT id id) id
+    , functionEq mkT runT p (firstT id) id
+    , functionEq mkT runT p (secondT id) id
+    , functionEq mkT runT p (bimapT f g) (firstT f . secondT g)
+    ]
```
Comment (tranma) [inline test/Test/P/Bifunctor/Trans.hs]: wow, this is beautiful

**Comment by tranma** [comment]
@jystic the only other semi-useful instance I could think of is `IdentityT` as `icicle` uses it (in the fresh name source I think?). Perhaps it's worth adding? ¯\_(ツ)_/¯ State/Writer/RWST/etc. are not possible as not all arguments are covariant.

this is cool! :+1:

**Comment by markhibberd** [comment]
:+1:

**Comment by tranma** [comment]
pls disregard my comment, `IdentityT` is over a functor, not a bifunctor

LGTM :postal_horn:

**Pull request merged**